torch/csrc/dynamo/guards.cpp

```diff
@@ -574,6 +574,371 @@ static struct PyModuleDef _module = {
+    return check_with_debug_info(value).first;
+  }
+
+  std::pair<bool, GuardDebugInfo> check_with_debug_info(py::object value) {
+    const std::pair<bool, GuardDebugInfo> result_pair = run_guards(value);
+    if (result_pair.first == false) {
+      _fail_count += 1;
+    }
+    return result_pair;
+  }
+
+  std::pair<bool, GuardDebugInfo> run_guards(py::object value) {
+    int debug_num_guards_executed = 0;
+    bool result = true;
+
+    // Iterate over leaf guards
+    for (const auto& guard : _leaf_guards) {
+      const std::pair<bool, GuardDebugInfo>& tmp =
+          guard->check_with_debug_info(value);
+      result &= tmp.first;
+      debug_num_guards_executed++;
+      // TODO (janimesh): Does this check adds overhead?
+      if (result == false) {
+        auto& debug_info = tmp.second;
+        return std::make_pair(
+            result,
+            GuardDebugInfo(
+                debug_num_guards_executed, debug_info.failure_reason));
+      }
+    }
+
+    // Iterate over accessors
+    std::string reason = "";
+    bool failed_on_first = true;
+    for (const auto& accessor : _accessors) {
+      auto& manager = accessor->get_guard_manager();
+      const std::pair<bool, GuardDebugInfo>& tmp =
+          manager->check_with_debug_info(accessor->access(value));
+      result &= tmp.first;
+      auto& debug_info = tmp.second;
+      debug_num_guards_executed += debug_info.num_guards_executed;
+      if (result == false) {
+        reason = debug_info.failure_reason;
+        break;
+      }
+      failed_on_first = false;
+    }
+
+    // failed_on_first is just an optimization to avoid sorting if we are
+    // failing on the first accessor itself. This is helpful when we have
+    // already sorted the guards once, and dont need to sort again.
+    if (result == false and failed_on_first == false) {
+      // Inplace sort the child guards by fail count. This moves the guard with
+      // higher fail count earlier in the queue, and enables fail fast for the
+      // next check_with_debug_info.
+
+      // An alternate implementation was to use priority queue directly on
+      // _accessors, but it was rejected because of the complexity of
+      // popping and creating a new pq on each run_guards. Moreover, this sort
+      // is happening on the unhappy path when check_with_debug_info guard
+      // fails. So, its probably ok.
+      std::sort(
+          _accessors.begin(),
+          _accessors.end(),
+          [](const std::unique_ptr<GuardAccessor>& a,
+             const std::unique_ptr<GuardAccessor>& b) {
+            return a->get_guard_manager()->fail_count() >=
+                b->get_guard_manager()->fail_count();
+          });
+    }
+    return std::make_pair(
+        result, GuardDebugInfo(debug_num_guards_executed, reason));
+  }
+
+  int fail_count() const {
+    return _fail_count;
+  }
+
+  // Returning raw pointers because we can't return unique_ptr and pybind does
```

Comment:
I don't get it, this is ownership ambiguity. Returning unique_ptr would mean you are passing ownership. If you have firm guarantees about lifecycle, return const refs?